src/BugsnagUnity/Delivery.cs

```diff
@@ -0,0 +1,172 @@
+        }
+
+        public void Send(IPayload payload)
+        {
+            if (payload.PayloadType == PayloadType.Event)
+            {
+                var report = (Report)payload;
+
+                lock (_callbackLock)
+                {
+                    foreach (var onSendErrorCallback in _configuration.GetOnSendErrorCallbacks())
+                    {
+                        try
+                        {
+                            if (!onSendErrorCallback.Invoke(report.Event))
+                            {
+                                return;
+                            }
+                        }
+                        catch
+                        {
+                            // If the callback causes an exception, ignore it and execute the next one
+                        }
+                    }
+                }
+                report.Event.RedactMetadata(_configuration);
+                report.ApplyEventPayload();
+            }
+            SerializeAndDeliverPayload(payload);
+        }
+
+        IEnumerator PushToServer(IPayload payload, byte[] body)
+        {
+            using (var req = new UnityWebRequest(payload.Endpoint.ToString()))
+            {
+                req.SetRequestHeader("Content-Type", "application/json");
+                req.SetRequestHeader("Bugsnag-Sent-At", DateTimeOffset.Now.ToString("o", CultureInfo.InvariantCulture));
+                foreach (var header in payload.Headers)
+                {
+                    req.SetRequestHeader(header.Key, header.Value);
+                }
+                req.uploadHandler = new UploadHandlerRaw(body);
+                req.downloadHandler = new DownloadHandlerBuffer();
+                req.method = UnityWebRequest.kHttpVerbPOST;
+                yield return req.SendWebRequest();
+
+                while (!req.isDone)
+                {
+                    yield return new WaitForEndOfFrame();
+                }
+                var code = req.responseCode;
+                if (code == 200 || code == 202)
+                {
+                    // success!
+                    FileManager.PayloadSendSuccess(payload);
+                }
+                else if (code >= 500 || code == 408 || code == 429 || req.isNetworkError || code < 400)
+                {
+                    // sending failed, cache payload to disk
+                    FileManager.SendPayloadFailed(payload);
+                }
+                _finishedCacheDeliveries.Add(payload.Id);
+            }
+        }
+
+        public void TrySendingCachedPayloads()
+        {
+            if (_cacheDeliveryInProcess)
+            {
+                return;
+            }
+            _cacheDeliveryInProcess = true;
+            try
+            {
+                _finishedCacheDeliveries.Clear();
+                var payloads = FileManager.GetCachedPayloadPaths();
+                MainThreadDispatchBehaviour.Instance().Enqueue(DeliverCachedPayloads(payloads));
```

Comment:
The other use of `MainThreadDispatchBehaviour` calls `CreateDispatchBehaviour` - do we need the equivalent here? Do these two separate calls to the dispatcher play nicely together?

Reply:
Added the check for it's existence 👍